### PR TITLE
Basic Backtranscludes filter

### PR DESCRIPTION
**Pull request by linonetwo**
## Summary

Feature "Adding backtranscludes filter"

## Purpose of change

This fixes https://github.com/Jermolene/TiddlyWiki5/issues/847

Add filters and indexers similar to links.

## Describe the solution

I basically copy&paste the code and replace "link" to "transclude" :P

## Testing

Copy&paste the test from links...

~~But I don't know how to run tests locally, there is no npm script to run `npm run test` for.~~

I see tests are passing in the CI.

## Additional context

Demo: https://tiddlywiki5-6ivxvwmuq-jermolene.vercel.app/#transcludes%20Operator%20(Examples):%5B%5Btranscludes%20Operator%20(Examples)%5D%5D%20%5B%5Btranscludes%20Operator%5D%5D%20%5B%5Bbacktranscludes%20Operator%20(Examples)%5D%5D%20%5B%5Bbacktranscludes%20Operator%5D%5D

This can also closes https://github.com/Jermolene/TiddlyWiki5/pull/4766

Example screenshots:

截屏2021-10-02 下午6 52 42
截屏2021-10-02 下午6 51 56

With this widget:

```js
<$list filter="[all[current]backtranscludes[]first[]sort[title]]">
	<label style="margin-left:auto;margin-right:auto;width:fit-content;display:block;">Transclusion (other)-{}->(this)</label>
</$list>

<<inverse-link-and-folder_list-links filter:"[all[current]backtranscludes[]]" class:"horizontal-list-links" >>

<$list filter="[all[current]backlinks[]first[]sort[title]]">
	<label style="margin-left:auto;margin-right:auto;width:fit-content;display:block;">Link (other)-[]->(this)</label>
</$list>

<<inverse-link-and-folder_list-links filter:"[all[current]backlinks[]]" class:"horizontal-list-links" >>
```

P.S. I'm try using PR template from https://github.com/CleverRaven/Cataclysm-DDA/blob/master/.github/pull_request_template.md , they are generating formated release note based on Summary section of PR.

### CODE REVIEW

**Comment by linonetwo** [comment]
> divide the tiddler info "References" tab into subtabs for "All", "Links" and "Transclusions".

This is somehow hard for me... So skit this time, hope someone more familiar with tab and translation can help me with this.

And I have trouble writing examples for transclusion, because the https://tiddlywiki.com/ seems to use zero transclusion...

**Comment by pmario** [comment]
In "[IDEA] Some way to click on a transcluded tiddler in draft mode" #5378  you mentioned that `{{title||template}}` won't work with this PR. ... 

If that's the case IMO the PR has to be fixed.

**Comment by linonetwo** [comment]
Yes, I see Transcluding via a template `{{Index||$:/core/ui/TagTemplate}}` is not shown in backtransclusions[].

But should it? It is not using transclusion widget, so `parseTreeNode.type === "transclude"` can't see it.

**Comment by pmario** [comment]
The parser output for `{{Index||$:/core/ui/TagTemplate}}` is: 

```
[
    {
        "type": "tiddler",
        "attributes": {
            "tiddler": {
                "type": "string",
                "value": "Index"
            }
        },
        "isBlock": true,
        "children": [
            {
                "type": "transclude",
                "attributes": {
                    "tiddler": {
                        "type": "string",
                        "value": "$:/core/ui/TagTemplate"
                    }
                },
                "isBlock": true
            }
        ]
    }
]
```

So the info is there.

**Comment by linonetwo** [comment]
I see...So it is actually transcluding `$:/core/ui/TagTemplate`

截屏2021-10-04 00 19 24

So maybe I should found this node's parent, and get `"type": "tiddler",` to find `Index` tiddler?

**Comment by pmario** [comment]
>So maybe I should found this node's parent, and get "type": "tiddler", to find Index tiddler?

Yes. I think the test has to be extended, that it first looks for node.type==="tiddler" and node.type==="transclude". If it finds the later one it can execute as it is now. 

If it finds node.type==="tiddler" it also has to check the 1st child. If it contains node.type==="transclude" it needs to use the `tiddler attribute` as a "backlink" info

**Comment by pmario** [comment]
> Will that make backtransclude impure?

No -- it will make it less performant. So we have to have a closer look at the code structure, to keep it fast

**Comment by pmario** [comment]
@linonetwo ... Did you have a closer look at PR #6666 ... IMO it may influence this PR. There are much more possibilities for transclusions now.

@Jermolene ... Did you miss this PR? You did not comment it. It seems to fix a long standing issue #847 that you created 2014 :)

**Comment by linonetwo** [comment]
This filter basically works, and maybe we can add more feature to it later...

**Comment by Jermolene** [comment]
Thanks for your patience @linonetwo.

The problem with this implementation is that it duplicates a lot of the code concerned with tracking links and backlinks; essentially it's the same code but looking for a different parse tree node. Given the similarities, I'm hoping that they can share implementation, using parameterised helper functions into which one would pass the details of the parse tree node.

It may also be useful to extend this to other parse tree nodes. For example, it might be helpful to offer a filter that is able to track image references so that one can display a list of the tiddlers that reference an image.

**Comment by saqimtiaz** [comment]
> It may also be useful to extend this to other parse tree nodes. For example, it might be helpful to offer a filter that is able to track image references so that one can display a list of the tiddlers that reference an image.

This would be very welcome. There is also the use case of being able to find all the images referenced in a tiddler that also involves searching through the parse tree nodes, for which I currently use a custom filter.

**Comment by linonetwo** [comment]
> using parameterised helper functions into which one would pass the details of the parse tree node

This is the new feature recently merged? I will take time to learn that.

I guess this will allow plugin authors tracking macro calls without modifying the core code.

**Comment by Jermolene** [comment]
> This is the new feature recently merged? I will take time to learn that.

No, this is nothing to do with the parameterised transclusion work in #6666. I was referring to the possibility that instead of having both a backtranscludes-indexer and a backlinks-indexer, that one could write a generic indexer that took as a parameter the details of the node that it is to target, and then use that to implement both the backtranscludes and backlinks indexers.

> TBO, I skim through the https://talk.tiddlywiki.org/t/parameterised-transclusion-introduction/4925 , but didn't immediately get how to use it. I will try again, just like how I come back to learn basic filter expression last year...

Feel free to ask questions there or over at #6666

> I guess this will allow plugin authors tracking macro calls without modifying the core code.

I'm not sure what you mean. Perhaps you're referring to the ability in #6666 to override built-in widgets?

**Comment by linonetwo** [comment]
> instead of having both a backtranscludes-indexer and a backlinks-indexer, that one could write a generic indexer

@Jermolene I've done the refactor to merge them. But things in $tw.wiki don't seem to be able to merge.

P.S. I'm not sure if we still need `(function(){` wrapper.

**Comment by linonetwo** [comment]
Demo: https://tiddlywiki5-6ivxvwmuq-jermolene.vercel.app/#transcludes%20Operator%20(Examples):%5B%5Btranscludes%20Operator%20(Examples)%5D%5D%20%5B%5Btranscludes%20Operator%5D%5D%20%5B%5Bbacktranscludes%20Operator%20(Examples)%5D%5D%20%5B%5Bbacktranscludes%20Operator%5D%5D

File: core/modules/wiki.js [hunk 1]
```diff
@@ -534,8 +534,8 @@ Return an array of tiddler titles that link to the specified tiddler
 */
 exports.getTiddlerBacklinks = function(targetTitle) {
 	var self = this,
-		backlinksIndexer = this.getIndexer("BacklinksIndexer"),
-		backlinks = backlinksIndexer && backlinksIndexer.lookup(targetTitle);
+		backIndexer = this.getIndexer("BackIndexer"),
+		backlinks = backIndexer && backIndexer.subIndexers.link.lookup(targetTitle);
 
 	if(!backlinks) {
```
Comment (linonetwo) [inline core/modules/wiki.js:540]: this `if` seems unnecessary, should I delete it?

**Comment by linonetwo** [comment]
@Jermolene Could you take the time to review this? So I can continue to implement section backlink after #7744 is merged too.

**Comment by pmario** [comment]
@Jermolene ... It would be nice to review this PR again too.

**Comment by linonetwo** [comment]
If we can confirm this new indexer design, I can start trying to cache the indexer in SQLite in https://github.com/tiddly-gittly/TidGi-Mobile , I want to make the startup time of mobile app down to 1s. It is too slow for my wiki.

**Comment by Jermolene** [comment]
Thank you for your patience, @linonetwo. This is an excellent PR, and I appreciate you making the changes that I suggested.

**Comment by kookma** [comment]
Thank you all.

Shouldn't indicate  `<<.from-version 5.3.3>>` in the doc?

**Comment by linonetwo** [comment]
This basically means ignore input title from the output, but backlinks[] will return it

图片

Are we going to include or ignore it? I can make it consistant in each transclusion case.

**Pull request merged**